beacon-chain/cache/depositsnapshot/merkle_tree.go

```diff
@@ -11,9 +15,185 @@ type MerkleTreeNode interface {
 	// IsFull returns whether there is space left for deposits.
 	IsFull() bool
 	// Finalize marks deposits of the Merkle tree as finalized.
-	Finalize(deposits uint, depth uint) MerkleTreeNode
+	Finalize(deposits uint64, depth uint64) MerkleTreeNode
 	// GetFinalized returns a list of hashes of all the finalized nodes and the number of deposits.
-	GetFinalized(result [][32]byte) ([][32]byte, uint)
+	GetFinalized(result [][32]byte) ([][32]byte, uint64)
 	// PushLeaf adds a new leaf node at the next available Zero node.
-	PushLeaf(leaf [32]byte, deposits uint, depth uint) MerkleTreeNode
+	PushLeaf(leaf [32]byte, depth uint64) MerkleTreeNode
+}
+
+func create(leaves [32]byte, depth uint64) MerkleTreeNode {
+	if leaves == [32]byte{0} {
+		return &ZeroNode{depth: depth}
+	}
+	if depth == 0 {
+		return &LeafNode{hash: [32]byte{leaves[0]}}
+	}
+	split := math.Min(math.PowerOf2(depth-1), uint64(len(leaves)))
+	left := create(leaves[0:split], depth-1)
+	right := create(leaves[split:], depth-1)
+	return &InnerNode{left: left, right: right}
+}
+
+func fromSnapshotParts(finalized [][32]byte, deposits uint64, level uint64) MerkleTreeNode {
+	if len(finalized) < 1 || deposits == 0 {
+		return &ZeroNode{
+			depth: level,
+		}
+	}
+	if deposits == math.PowerOf2(level) {
+		return &FinalizedNode{
+			depositCount: deposits,
+			hash:         finalized[0],
+		}
+	}
+	node := InnerNode{}
+	if leftSubtree := math.PowerOf2(level - 1); deposits <= leftSubtree {
+		node.left = fromSnapshotParts(finalized, deposits, level-1)
+		node.right = &ZeroNode{depth: level - 1}
+
+	} else {
+		node.left = &FinalizedNode{
+			depositCount: leftSubtree,
+			hash:         finalized[0],
+		}
+		node.right = fromSnapshotParts(finalized[1:], deposits-leftSubtree, level-1)
+	}
+	return &node
+}
+
+func generateProof(tree InnerNode, index uint64, depth uint64) ([32]byte, [32]byte) {
+	var proof [32]byte
+	node := tree
+	for depth > 0 {
+		ithBit := (index >> (depth - 1)) & 0x1
+		if ithBit == 1 {
+			proof = append(proof[:], node.left.GetRoot())
+			node = node.right
+		} else {
+			proof = append(proof[:], node.right.GetRoot())
+			node = node.left
+		}
+		depth -= 1
+	}
+	//	TODO Add reverse
```

Comment:
Good opportunity to use generics! https://github.com/faiface/generics/blob/8cf65f0b43803410724d8c671cb4d328543ba07d/examples/sliceutils/sliceutils.go#L9

Reply:
Nice! Thank you